vulnerabilities/import_runner.py

```diff
@@ -0,0 +1,304 @@
+        shared across many packages, from the same data source and from different data sources. For example, a
+        vulnerability in the Linux kernel is mentioned by advisories from all Linux distributions that package this
+        kernel version.
+        """
+        logger.debug(f'Starting import for {self.importer.name}.')
+        data_source = self.importer.make_data_source(cutoff_date=cutoff_date, batch_size=self.batch_size)
+
+        with data_source as ds:
+            _process_new_advisories(ds)
+            _process_updated_advisories(ds)
+
+        self.importer.last_run = datetime.datetime.utcnow()
+        self.importer.save()
+
+        logger.debug(f'Successfully finished import for {self.importer.name}.')
+
+
+def _process_new_advisories(data_source):
+    for batch in data_source.added_advisories():
+        impacted, resolved = _collect_purls(batch)
+        impacted, resolved = _bulk_insert_packages(impacted, resolved)
+
+        vulnerabilities = _insert_vulnerabilities_and_references(batch)
+
+        _bulk_insert_impacted_and_resolved_packages(batch, vulnerabilities, impacted, resolved)
+
+
+def _process_updated_advisories(data_source):
+    """
+    TODO: Make efficient; Current implementation needs way too many DB queries.
+    """
+    for batch in data_source.updated_advisories():
+        for advisory in batch:
+            vuln, _ = _get_or_create_vulnerability(advisory)
+
+            for url in advisory.references:
+                models.VulnerabilityReference.objects.get_or_create(vulnerability=vuln, url=url)
+
+            for ipkg_url in advisory.impacted_package_urls:
+                pkg, created = _get_or_create_package(ipkg_url)
+
+                # FIXME Does not work yet due to cascading deletes.
+                # if not created:
+                #     qs = models.ResolvedPackage.objects.filter(vulnerability_id=vuln.id, package_id=pkg.id)
+                #     if qs:
+                #         qs[0].delete()
+
+                models.ImpactedPackage.objects.get_or_create(vulnerability_id=vuln.id, package_id=pkg.id)
+
+            for rpkg_url in advisory.resolved_package_urls:
+                pkg, created = _get_or_create_package(rpkg_url)
+
+                # FIXME Does not work yet due to cascading deletes.
+                # if not created:
+                #     qs = models.ImpactedPackage.objects.filter(vulnerability_id=vuln.id, package_id=pkg.id)
+                #     if qs:
+                #         qs[0].delete()
+
+                models.ResolvedPackage.objects.get_or_create(vulnerability_id=vuln.id, package_id=pkg.id)
+
+
+def _get_or_create_vulnerability(advisory: Advisory) -> Tuple[models.Vulnerability, bool]:
+    query_kwargs = {}
+
+    if advisory.cve_id:
+        query_kwargs['cve_id'] = advisory.cve_id
+    else:
+        query_kwargs = {'summary': advisory.summary}
+    vuln, created = models.Vulnerability.objects.get_or_create(**query_kwargs)
```

Comment:
Question: Is there any reason, for not considering the case where `cve_id` and `summary` both are available, in such case shouldn' those be added to `query_kwargs`?

 By the way, this is really great work.

Reply:
Thanks for the kind words! It sure is taking long enough. :)

Including both in the query would only work if we add a "unique together" constraint for `cve_id` and `summary`. As it is now, if we have the vulnerability with a different summary and include both in the call to `get_or_create()`, we get this:
```
In [54]: Vulnerability.objects.count()                                                                                                      
Out[54]: 0

In [55]: Vulnerability.objects.create(cve_id='abc', summary='remote root many bad')                                                         
Out[55]: <Vulnerability: remote root many bad>

In [56]: Vulnerability.objects.get_or_create(cve_id='abc', summary='other summary for the same vuln')                                       
---------------------------------------------------------------------------
DoesNotExist                              Traceback (most recent call last)
~/code/vulnerablecode/venv/lib/python3.8/site-packages/django/db/models/query.py in get_or_create(self, defaults, **kwargs)
    558         try:
--> 559             return self.get(**kwargs), False
    560         except self.model.DoesNotExist:

~/code/vulnerablecode/venv/lib/python3.8/site-packages/django/db/models/query.py in get(self, *args, **kwargs)
    414         if not num:
--> 415             raise self.model.DoesNotExist(
    416                 "%s matching query does not exist." %

DoesNotExist: Vulnerability matching query does not exist.

During handling of the above exception, another exception occurred:

UniqueViolation                           Traceback (most recent call last)
~/code/vulnerablecode/venv/lib/python3.8/site-packages/django/db/backends/utils.py in _execute(self, sql, params, *ignored_wrapper_args)
     85             else:
---> 86                 return self.cursor.execute(sql, params)
     87 

UniqueViolation: duplicate key value violates unique constraint "vulnerabilities_vulnerability_cve_id_key"
DETAIL:  Key (cve_id)=(abc) already exists.


The above exception was the direct cause of the following exception:

IntegrityError                            Traceback (most recent call last)
<ipython-input-56-196121bcc8ce> in <module>
----> 1 Vulnerability.objects.get_or_create(cve_id='abc', summary='other summary for the same vuln')

~/code/vulnerablecode/venv/lib/python3.8/site-packages/django/db/models/manager.py in manager_method(self, *args, **kwargs)
     80         def create_method(name, method):
     81             def manager_method(self, *args, **kwargs):
---> 82                 return getattr(self.get_queryset(), name)(*args, **kwargs)
     83             manager_method.__name__ = method.__name__
     84             manager_method.__doc__ = method.__doc__

--------------- preposterously long traceback snipped ----------------------

~/code/vulnerablecode/venv/lib/python3.8/site-packages/django/db/backends/utils.py in _execute(self, sql, params, *ignored_wrapper_args)
     84                 return self.cursor.execute(sql)
     85             else:
---> 86                 return self.cursor.execute(sql, params)
     87 
     88     def _executemany(self, sql, param_list, *ignored_wrapper_args):

IntegrityError: duplicate key value violates unique constraint "vulnerabilities_vulnerability_cve_id_key"
DETAIL:  Key (cve_id)=(abc) already exists.
```

NB: As you see further below in the code I made the completely arbitrary decision to update the database with the summary from the currently imported data source in the case they differ.